src/AvaloniaEdit/Rendering/BackgroundGeometryBuilder.cs

```diff
@@ -260,10 +260,13 @@ private static IEnumerable<Rect> ProcessTextLines(TextView textView, VisualLine
 					} else {
 						right = visualLine.GetTextLineVisualXPosition(lastTextLine, segmentEndVc);
 					}
+
+					left -= scrollOffset.X;
+					right -= scrollOffset.X;
 					Rect extendSelection = new Rect(Math.Min(left, right), y, Math.Abs(right - left), line.Height);
 					if (lastRect != default) {
 						if (extendSelection.Intersects(lastRect)) {
-							lastRect.Union(extendSelection);
+							lastRect = lastRect.Union(extendSelection);
```

Comment:
This is a second, unrelated fix where the selection is not properly extended because `Union` is use wrong.